<?xml version="1.0"?>
<!--***********************************************************
	IHMC Interface Node
	Emily Sheetz, NSTGRO VTE 2021

	run either:
		$ roslaunch IHMCMsgInterface ihmc_interface_node.launch
		OR
		$ rosrun IHMCMsgInterface ihmc_interface_node.py
************************************************************-->
<launch>
	<arg name="controllers" default="true"/> <!-- indicates if joint commands come from controllers; will change queueing properties of IHMC messages -->
	<arg name="managing_node" default="IHMCInterfaceNode"/> <!-- only necessary if controllers flag is true --> <!-- "ControllerTestNode" -->

	<arg name="launch_footstep_services" default="false"/> <!-- indicates if planning and executing services should be launched -->

	<!-- if not testing with controllers, robot pose and joint states will come from IKModuleTestNode -->
	<arg unless="$(arg controllers)" name="pelvis_tf_topic" value="/IKModuleTestNode/nstgro20_valkyrie_ik/pelvis_transform"/>
	<arg unless="$(arg controllers)" name="joint_command_topic" value="/IKModuleTestNode/nstgro20_valkyrie_ik/joint_commands"/>
	<!-- if testing with controllers, robot pose and joint states will come from managing_node -->
	<arg if="$(arg controllers)" name="pelvis_tf_topic" value="controllers/output/ihmc/pelvis_transform"/>
	<arg if="$(arg controllers)" name="joint_command_topic" value="controllers/output/ihmc/joint_commands"/>
	<arg if="$(arg controllers)" name="status_topic" value="controllers/output/ihmc/controller_status"/>
	<arg if="$(arg controllers)" name="controlled_link_topic" value="controllers/output/ihmc/controlled_link_ids"/>
	<arg if="$(arg controllers)" name="hand_pose_command_topic" value="controllers/output/ihmc/cartesian_hand_targets"/>
	<arg if="$(arg controllers)" name="receive_cartesian_goals_topic" value="controllers/output/ihmc/receive_cartesian_goals"/>
	<arg if="$(arg controllers)" name="moveit_traj_topic" value="moveit_planned_robot_trajectory"/>
	<arg if="$(arg controllers)" name="receive_moveit_traj_topic" value="receive_moveit_trajectories"/>

	<arg name="debug" default="false"/>
	<arg unless="$(arg debug)" name="launch_prefix" value="" />
	<arg if="$(arg debug)" name="launch_prefix" value="xterm -e gdb **args" />

	<group if="$(arg launch_footstep_services)">
		<include file="$(find val_footstep_planner_executor)/launch/footstep_planner_executor_server_node.launch"/>
	</group>

	<node launch-prefix="$(arg launch_prefix)" pkg="ihmc_msg_interface" type="ihmc_interface_node" name="IHMCInterfaceNode" output="screen">
		<param name="commands_from_controllers" value="$(arg controllers)"/>
		<param name="joint_command_topic" value="$(arg joint_command_topic)"/>
		<param name="pelvis_tf_topic" value="$(arg pelvis_tf_topic)"/>
		<!-- only controllers will send statuses and controlled links, otherwise status topic not needed -->
		<param if="$(arg controllers)" name="managing_node" value="$(arg managing_node)"/>
		<param if="$(arg controllers)" name="status_topic" value="$(arg status_topic)"/>
		<param if="$(arg controllers)" name="controlled_link_topic" value="$(arg controlled_link_topic)"/>
		<param if="$(arg controllers)" name="hand_pose_command_topic" value="$(arg hand_pose_command_topic)"/>
		<param if="$(arg controllers)" name="receive_cartesian_goals_topic" value="$(arg receive_cartesian_goals_topic)"/>
		<param if="$(arg controllers)" name="moveit_traj_topic" value="$(arg moveit_traj_topic)"/>
		<param if="$(arg controllers)" name="receive_moveit_traj_topic" value="$(arg receive_moveit_traj_topic)"/>
		<!--<param name="" type="" value=""/> -->
	</node>
</launch>
